internal/graph/check_test.go

```diff
@@ -3072,3 +3076,990 @@ func TestConsumeUsersets(t *testing.T) {
+						},
+					},
+				},
+			},
+		},
+		{
+			name:         "shortcut_with_error_at_end",
+			poolSize:     1,
+			ctxCancelled: false,
+			dispatchMsgs: []dispatchMsg{
+				{
+					shortCircuit: true,
+				},
+				{
+					err: fmt.Errorf("should_not_process_this"),
+				},
+			},
+			expectedOutcomes: []checkOutcome{
+				{
+					resp: &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: 0,
+						},
+					},
+				},
+			},
+		},
+		{
+			name:         "multiple_dispatches",
+			poolSize:     1,
+			ctxCancelled: false,
+			dispatchMsgs: []dispatchMsg{
+				{
+					dispatchParams: &dispatchParams{
+						parentReq: req,
+						tk:        tuple.NewTupleKey("group:1", "member", "user:maria"),
+					},
+				},
+				{
+					dispatchParams: &dispatchParams{
+						parentReq: req,
+						tk:        tuple.NewTupleKey("group:2", "member", "user:maria"),
+					},
+				},
+			},
+			mockedDispatchResponse: []*ResolveCheckResponse{
+				{
+					Allowed: true,
+					ResolutionMetadata: &ResolveCheckResponseMetadata{
+						DatastoreQueryCount: datastoreQueryCount,
+					},
+				},
+				{
+					Allowed: false,
+					ResolutionMetadata: &ResolveCheckResponseMetadata{
+						DatastoreQueryCount: datastoreQueryCount,
+					},
+				},
+			},
+			expectedOutcomes: []checkOutcome{
+				{
+					resp: &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: datastoreQueryCount,
+						},
+					},
+				},
+				{
+					resp: &ResolveCheckResponse{
+						Allowed: false,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: datastoreQueryCount,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			dutPool := concurrency.NewPool(context.Background(), tt.poolSize)
+			ctx := context.Background()
+			var cancel context.CancelFunc
+			if tt.ctxCancelled {
+				ctx, cancel = context.WithCancel(ctx)
+				cancel()
+			}
+
+			checker := NewLocalChecker()
+			defer checker.Close()
+			mockResolver := NewMockCheckResolver(ctrl)
+			checker.SetDelegate(mockResolver)
+
+			for _, mockedDispatchResponse := range tt.mockedDispatchResponse {
+				mockResolver.EXPECT().ResolveCheck(gomock.Any(), gomock.Any()).Times(1).Return(mockedDispatchResponse, nil)
+			}
+
+			dispatchMsgChan := make(chan dispatchMsg, 100)
+			for _, dispatchMsg := range tt.dispatchMsgs {
+				dispatchMsgChan <- dispatchMsg
+			}
+			outcomeChan := make(chan checkOutcome, 100)
+
+			go checker.processDispatches(ctx, dutPool, dispatchMsgChan, outcomeChan)
+
+			// now, close the channel to simulate everything is sent
+			close(dispatchMsgChan)
+			outcomes := helperReceivedOutcome(outcomeChan)
+
+			require.Equal(t, tt.expectedOutcomes, outcomes)
```

Comment:
Based on my other suggestion, this can be simplified:

```suggestion
			dispatchMsgChan := make(chan dispatchMsg, 100)
			for _, dispatchMsg := range tt.dispatchMsgs {
				dispatchMsgChan <- dispatchMsg
			}

			outcomeChan := checker.processDispatches(ctx, uint32(tt.poolSize), dispatchMsgChan)

			// now, close the channel to simulate everything is sent
			close(dispatchMsgChan)
			outcomes := helperReceivedOutcome(outcomeChan)

			require.Equal(t, tt.expectedOutcomes, outcomes)
```

and you can remove this line
- `dutPool := concurrency.NewPool(context.Background(), tt.poolSize)`